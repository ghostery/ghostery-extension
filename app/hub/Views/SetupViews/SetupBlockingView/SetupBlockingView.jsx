/**
 * Setup Blocking View Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';

/**
 * A Functional React component for rendering the Setup Blocking View
 * @return {JSX} JSX for rendering the Setup Blocking View of the Hub app
 * @memberof HubComponents
 */
const SetupBlockingView = props => (
	<div className="row align-center">
		<div className="columns small-12 large-10">
			<div className="SetupBlocking">
				<div className="row small-up-4">
					{props.choices.map((choice) => {
						const choiceSelected = choice.name === props.blockingPolicy;
						const bigCheckSrc = '/app/images/hub/setup/block-selected.svg';
						const choiceBoxClassNames = ClassNames('clickable', 'flex-container', 'flex-dir-column', {
							SetupBlocking__choiceBox: true,
							'SetupBlocking--selected': choiceSelected,
						});

						return (
							<div key={`block-value-${choice.name}`} className="columns">
								<div>
									<label
										htmlFor={`input-block-${choice.name}`}
										className={choiceBoxClassNames}
										onClick={() => {
											if (choice.name === 'BLOCKING_POLICY_CUSTOM') {
												props.handleCustomClick();
											}
										}}
									>
										<div className="SetupBlocking__imageContainer flex-container align-center-middle">
											<img src={choiceSelected ? bigCheckSrc : choice.image} />
										</div>
										<div className="SetupBlocking__textContainer flex-child-grow flex-container align-center-middle">
											{choice.text}
										</div>
									</label>
									<div className="SetupBlocking__description">{choice.description}</div>
								</div>

								<input
									type="radio"
									name={choice.name}
									value={choice.name}
									id={`input-block-${choice.name}`}
									checked={props.blockingPolicy === choice.name}
									onChange={props.handleSelection}
								/>
							</div>
						);
					})}
				</div>
			</div>
		</div>
	</div>
);

// PropTypes ensure we pass required props of the correct type
SetupBlockingView.propTypes = {
	blockingPolicy: PropTypes.string.isRequired,
	handleSelection: PropTypes.func.isRequired,
	choices: PropTypes.arrayOf(PropTypes.shape({
		name: PropTypes.string.isRequired,
		image: PropTypes.string.isRequired,
		text: PropTypes.string.isRequired,
		description: PropTypes.string.isRequired,
	})).isRequired,
};

export default SetupBlockingView;
