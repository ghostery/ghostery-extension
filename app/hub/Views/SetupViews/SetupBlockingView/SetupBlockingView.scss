/**
 * Setup Blocking View Sass
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

// Setup Blocking View
.SetupBlocking__description {
	display: none;
	max-width: 180px;
	margin-top: 30px;
	padding: 0 5px;
	color: #920094;
	font-size: 12px;
	line-height: 1.75;
	text-align: center;
}
.SetupBlocking__choiceBox {
	border: 2px solid rgba(#000000, 0);
	min-height: 180px;
	max-width: 180px;
}
.SetupBlocking--selected,
.SetupBlocking__choiceBox:hover {
	border-color: #930194;
}
.SetupBlocking--selected + .SetupBlocking__description,
.SetupBlocking__choiceBox:hover + .SetupBlocking__description {
	display: block;
}
.SetupBlocking input {
	display: none;
}
.SetupBlocking__imageContainer {
	height: 125px;
	background-color: #f7f7f7;
}
.SetupBlocking__imageContainer img {
	max-width: 75px;
	max-height: 75px;
}
.SetupBlocking__textContainer {
	font-size: 14px;
	font-weight: 500;
	line-height: 1.36;
	padding: 0 5px;
	text-align: center;
	color: #4a4a4a;
	background-color: #e7ecee;
}
