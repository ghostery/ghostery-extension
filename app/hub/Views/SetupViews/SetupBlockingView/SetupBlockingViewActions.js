/**
 * Setup Blocking View Action creators
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import { log, sendMessageInPromise } from '../../../utils';
import { SET_BLOCKING_POLICY } from '../../SetupView/SetupViewConstants';

export function setBlockingPolicy(actionData) {
	return function (dispatch) {
		return sendMessageInPromise(SET_BLOCKING_POLICY, actionData).then((data) => {
			dispatch({
				type: SET_BLOCKING_POLICY,
				data,
			});
		}).catch((err) => {
			log('setupBlocking Action setBlockingPolicy Error', err);
		});
	};
}
