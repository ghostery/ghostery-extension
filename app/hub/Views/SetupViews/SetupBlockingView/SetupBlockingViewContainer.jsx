/**
 * Setup Blocking View Container
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import SetupBlockingView from './SetupBlockingView';

/**
 * @class Implement the Setup Blocking View for the Ghostery Hub
 * @extends Component
 * @memberof HubComponents
 */
class SetupBlockingViewContainer extends Component {
	/**
	 * Lifecycle Event
	 */
	componentWillMount() {
		const title = t('hub_setup_page_title_blocking');
		window.document.title = title;

		const { index, setup, sendMountActions } = this.props;
		this.props.actions.setSetupNavigation({
			activeIndex: index,
			hrefPrev: false,
			hrefNext: `/setup/${index + 1}`,
			hrefDone: '/',
			textPrev: false,
			textNext: t('hub_setup_nav_next'),
			textDone: t('hub_setup_exit_flow'),
		});

		if (sendMountActions) {
			const { blockingPolicy } = setup;
			this.props.actions.setBlockingPolicy({ blockingPolicy });
		}
	}

	/**
	* Function to handle a change made on the Setup Blocking View
	* @param  {Object} event the click event
	*/
	_handleChange = (event) => {
		const blockingPolicy = event.target.value;
		this.props.actions.setBlockingPolicy({ blockingPolicy });
	}

	/**
	 * React's required render function. Returns JSX
	 * @return {JSX} JSX for rendering the Setup Blocking View of the Hub app
	 */
	render() {
		const { blockingPolicy } = this.props.setup;
		const choices = [
			{
				name: 'BLOCKING_POLICY_RECOMMENDED',
				image: '/app/images/hub/setup/block-recommended.svg',
				text: t('hub_setup_blocking_text_recommended'),
				description: t('hub_setup_blocking_description_recommended'),
			},
			{
				name: 'BLOCKING_POLICY_NOTHING',
				image: '/app/images/hub/setup/block-none.svg',
				text: t('hub_setup_blocking_text_nothing'),
				description: t('hub_setup_blocking_description_nothing'),
			},
			{
				name: 'BLOCKING_POLICY_EVERYTHING',
				image: '/app/images/hub/setup/block-all.svg',
				text: t('hub_setup_blocking_text_everything'),
				description: t('hub_setup_blocking_description_everything'),
			},
			{
				name: 'BLOCKING_POLICY_CUSTOM',
				image: '/app/images/hub/setup/block-custom.svg',
				text: t('hub_setup_blocking_text_custom'),
				description: t('hub_setup_blocking_description_custom'),
			},
		];

		return <SetupBlockingView blockingPolicy={blockingPolicy} choices={choices} handleSelection={this._handleChange} />;
	}
}

// PropTypes ensure we pass required props of the correct type
SetupBlockingViewContainer.propTypes = {
	index: PropTypes.number.isRequired,
	actions: PropTypes.shape({
		setSetupNavigation: PropTypes.func.isRequired,
		setBlockingPolicy: PropTypes.func.isRequired,
	}).isRequired,
};

export default SetupBlockingViewContainer;
