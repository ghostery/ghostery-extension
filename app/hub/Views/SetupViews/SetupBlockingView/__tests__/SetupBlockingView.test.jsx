/**
 * Setup Blocking View Test Component
 *
 * Ghostery Browser Extension
 * https://www.ghostery.com/
 *
 * Copyright 2018 Ghostery, Inc. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import SetupBlockingView from '../SetupBlockingView';

describe('app/hub/Views/SetupViews/SetupBlockingView component', () => {
	describe('Snapshot tests with react-test-renderer', () => {
		test('setup blocking view is rendered correctly', () => {
			const initialState = {
				blockingPolicy: 'test',
				choices: [
					{
						name: 'test',
						image: 'choice-test-image',
						text: 'This is your TEST choice',
						description: 'This is your TEST below text',
					},
					{
						name: 'example',
						image: 'choice-example-image',
						text: 'This is your EXAMPLE choice',
						description: 'This is your EXAMPLE below text',
					},
				],
				handleSelection: () => {},
			};

			const component = renderer.create(
				<SetupBlockingView {...initialState} />
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('More Snapshot tests with react-test-renderer, but for edge cases', () => {
		test('edge case where choices is an empty array', () => {
			const initialState = {
				blockingPolicy: 'test',
				choices: [],
				handleSelection: () => {},
			};

			const component = renderer.create(
				<SetupBlockingView {...initialState} />
			).toJSON();
			expect(component).toMatchSnapshot();
		});
	});

	describe('Shallow snapshot tests rendered with Enzyme', () => {
		test('the happy path of the component', () => {
			const initialState = {
				blockingPolicy: 'test',
				choices: [
					{
						name: 'test',
						image: 'choice-test-image',
						text: 'This is your TEST choice',
						description: 'This is your TEST below text',
					},
					{
						name: 'example',
						image: 'choice-example-image',
						text: 'This is your EXAMPLE choice',
						description: 'This is your EXAMPLE below text',
					},
				],
				handleSelection: () => {},
			};

			const component = shallow(<SetupBlockingView {...initialState} />);
			expect(component.find('.SetupBlocking').length).toBe(1);
			expect(component.find('.SetupBlocking__choiceBox').length).toBe(2);
			expect(component.find('.SetupBlocking--selected').length).toBe(1);
			expect(component.find('.SetupBlocking__imageContainer').length).toBe(2);
			expect(component.find('.SetupBlocking__textContainer').length).toBe(2);
		});

		test('the non-happy path of the component', () => {
			const initialState = {
				blockingPolicy: 'test',
				choices: [],
				handleSelection: () => {},
			};

			const component = shallow(<SetupBlockingView {...initialState} />);
			expect(component.find('.SetupBlocking').length).toBe(1);
			expect(component.find('.SetupBlocking__choiceBox').length).toBe(0);
			expect(component.find('.SetupBlocking--selected').length).toBe(0);
			expect(component.find('.SetupBlocking__imageContainer').length).toBe(0);
			expect(component.find('.SetupBlocking__textContainer').length).toBe(0);
		});
	});
});
