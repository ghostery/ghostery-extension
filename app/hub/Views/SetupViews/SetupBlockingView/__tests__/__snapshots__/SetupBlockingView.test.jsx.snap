// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/hub/Views/SetupViews/SetupBlockingView component More Snapshot tests with react-test-renderer, but for edge cases edge case where choices is an empty array 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12 large-10"
  >
    <div
      className="SetupBlocking"
    >
      <div
        className="row small-up-4"
      />
    </div>
  </div>
</div>
`;

exports[`app/hub/Views/SetupViews/SetupBlockingView component Snapshot tests with react-test-renderer setup blocking view is rendered correctly 1`] = `
<div
  className="row align-center"
>
  <div
    className="columns small-12 large-10"
  >
    <div
      className="SetupBlocking"
    >
      <div
        className="row small-up-4"
      >
        <div
          className="columns"
        >
          <div>
            <label
              className="clickable flex-container flex-dir-column SetupBlocking__choiceBox SetupBlocking--selected"
              htmlFor="input-block-test"
              onClick={[Function]}
            >
              <div
                className="SetupBlocking__imageContainer flex-container align-center-middle"
              >
                <img
                  src="/app/images/hub/setup/block-selected.svg"
                />
              </div>
              <div
                className="SetupBlocking__textContainer flex-child-grow flex-container align-center-middle"
              >
                This is your TEST choice
              </div>
            </label>
            <div
              className="SetupBlocking__description"
            >
              This is your TEST below text
            </div>
          </div>
          <input
            checked={true}
            id="input-block-test"
            name="test"
            onChange={[Function]}
            type="radio"
            value="test"
          />
        </div>
        <div
          className="columns"
        >
          <div>
            <label
              className="clickable flex-container flex-dir-column SetupBlocking__choiceBox"
              htmlFor="input-block-example"
              onClick={[Function]}
            >
              <div
                className="SetupBlocking__imageContainer flex-container align-center-middle"
              >
                <img
                  src="choice-example-image"
                />
              </div>
              <div
                className="SetupBlocking__textContainer flex-child-grow flex-container align-center-middle"
              >
                This is your EXAMPLE choice
              </div>
            </label>
            <div
              className="SetupBlocking__description"
            >
              This is your EXAMPLE below text
            </div>
          </div>
          <input
            checked={false}
            id="input-block-example"
            name="example"
            onChange={[Function]}
            type="radio"
            value="example"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
